README.md

```diff
@@ -1,16 +1,644 @@
+          "to": "bob@example.com",
+          "subject": "DNSLink for example.com",
+          "body": {"ucan/promise": ["/", "update-dns", "$http.body"]}
+        }
+      ]
+    },
+    "notify-carol": {
+      "using": "mailto://alice@example.com",
+      "do": "msg/send",
+      "inputs": [
+        {
+          "to": "carol@example.com",
+          "subject": "DNSLink for example.com",
+          "body": {"ucan/promise": ["/", "update-dns", "$http.body"]}
+        }
+      ]
+    },
+    "log-as-done": {
+      "using": "https://example.com/report"
+      "do": "crud/update"
+      "inputs": [
+        {
+          "from": "mailto://alice@exmaple.com",
+          "to": ["bob@exmaple.com", "carol@example.com"],
+          "event": "email-notification",
+        },
+        {
+          "_": {"ucan/promise": ["/", "notify-bob"]}
+        },
+        {
+          "_": {"ucan/promise": ["/", "notify-carol"]}
+        }
+      ]
+    }
+  },
+  "sig": "bdNVZn_uTrQ8bgq5LocO2y3gqIyuEtvYWRUH9YT-SRK6v_SX8bjt-VZ9JIPVTdxkWb6nhVKBt6JGpgnjABpOCA"
+}
+```
+
+### 5.2.2 Serial Pipeline
+
+```
+                ┌────────────────────────────┐
+                │                            │
+                │ dns://example.com?TYPE=TXT │
+                │        crud/update         │
+                │                            │
+                └───────┬──────────┬─────────┘
+                        │          │
+                        │          │
+                        │          │
+                        │      ┌───▼────────────────────────┐
+                        │      │                            │
+                        │      │ mailto://alice@example.com │
+                        │      │          msg/send          │
+                        │      │      carol@exmaple.com     │
+                        │      │                            │
+                        │      └───┬────────────────────────┘
+                        │          │
+┌───────────────────────┼──────────┼──────────┐
+│                       │          │          │
+│ ┌─────────────────────▼───┐      │          │
+│ │                         │      │          │
+│ │ mailto:alice@exaple.com │      │          │
+│ │         msg/send        │      │          │
+│ │     bob@example.com     │      │          │
+│ │                         │      │          │
+│ └─────────────────────┬───┘      │          │
+│                       │          │          │
+│                       │          │          │
+│              ┌────────▼──────────▼────────┐ │
+│              │                            │ │
+│              │ https://example.com/events │ │
+│              │         crud/create        │ │
+│              │                            │ │
+│              └────────────────────────────┘ │
+│                                             │
+└─────────────────────────────────────────────┘
+```
+
+ ``` json
+{
+  "ucan/invoke": "Qmd4trNUhgWwsBbSsYBEWJqgiHyLrnhZ8u1DJsWsEKeuuF",
```

Comment:
```suggestion
  "ucan/invoke": { "/": "Qmd4trNUhgWwsBbSsYBEWJqgiHyLrnhZ8u1DJsWsEKeuuF" },
```

Reply:
Totally get that this is how dag-json looks, but also it's also pretty noisy syntax when we know that these are CIDs. Is it possible to read an array of human-friendly CIDs as strings, or does the IPLD tooling help us here with the wrappers?

Reply:
IPLD tooling will transparently roundtrip CIDs and Bytes, but if you use strings for either IPLD will just treat them as strings which is why I think this is misleading.

Reply:
Not really misleading, but you need another pass to do the parsing of CIDs yourself, which is what I had to do for `prf` in UCAN for example. If they were encoded as `{ "/": "bafy...link" }` I would not have to pre/post process that data.

Reply:
> I would not have to pre/post process that data.

Ah, you mean from the existing IPLD tooling. So just to say this out loud: someone _could_ have a cleaner-looking JSON document, but they'd need to understand the type information of the fields and wrap them in the appropriate wrappers before feeding it into IPLD tools. Makes sense!

I may add text to that effect to the intro, and change the examples to look like your suggestions above 👍